bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/CorrectEnsemblePlacementCommand.java

```diff
@@ -0,0 +1,321 @@
+    public static class CorrectEnsemblePlacementFlags extends CliFlags {
+        @Parameter(names = { "-l", "--ledgerids" },
+                description = "Target ledger IDs to relocate. Multiple can be specified, comma separated.",
+                splitter = CommaParameterSplitter.class, required = true)
+        private List<Long> ledgerIds;
+
+        @Parameter(names = { "-dr", "--dryrun" },
+                description = "Printing the relocation plan w/o doing actual relocation")
+        private boolean dryRun;
+
+        @Parameter(names = { "-f", "--force" }, description = "Force relocation without confirmation")
+        private boolean force;
+
+        @Parameter(names = {"-sk", "--skipOpenLedgers"}, description = "Skip relocating open ledgers")
+        private boolean skipOpenLedgers;
+    }
+
+    @Override
+    public boolean apply(ServerConfiguration conf, CorrectEnsemblePlacementFlags flags) {
+        try {
+            if (flags.dryRun) {
+                LOG.info("The dry-run output could change every time you run"
+                        + " since the selection of bookies replaced includes some randomness.");
+            }
+            if (!flags.skipOpenLedgers) {
+                LOG.warn("Try to relocate also open ledgers. It is not recommended.");
+            }
+            try {
+                if (!flags.force) {
+                    final boolean confirm =
+                            IOUtils.confirmPrompt("Are you sure to relocate target ledgers?");
+                    if (!confirm) {
+                        LOG.error("Relocation is aborted.");
+                        return false;
+                    }
+                }
+            } catch (IOException e) {
+                LOG.error("Error during relocation", e);
+                return false;
+            }
+
+            final ClientConfiguration clientConf = new ClientConfiguration(conf);
+            final BookKeeper bookKeeper = new BookKeeper(clientConf);
+            final BookKeeperAdmin admin = new BookKeeperAdmin(clientConf);
+            return relocate(conf, flags, bookKeeper, admin);
+        } catch (Exception e) {
+            throw new UncheckedExecutionException(e.getMessage(), e);
+        }
+    }
+
+    @VisibleForTesting
+    public boolean relocate(ServerConfiguration conf, CorrectEnsemblePlacementFlags flags,
+                            BookKeeper bookKeeper, BookKeeperAdmin admin) throws Exception {
+        final EnsemblePlacementPolicy placementPolicy = bookKeeper.getPlacementPolicy();
+
+        @Cleanup
+        final MetadataBookieDriver metadataDriver = instantiateMetadataDriver(conf);
+        @Cleanup
+        final LedgerManagerFactory lmf = metadataDriver.getLedgerManagerFactory();
+        @Cleanup
+        final LedgerUnderreplicationManager lum = lmf.newLedgerUnderreplicationManager();
+
+        final List<Long> targetLedgers =
+                flags.ledgerIds.stream().parallel().distinct().filter(ledgerId -> {
```

Comment:
What's the reason for 'parallel'?
This loop is doing metadata operations, you can kill the zookeeper if you don't throttle properly the requests.

Reply:
> What's the reason for 'parallel'?

It was no strong reason.

> This loop is doing metadata operations, you can kill the zookeeper if you don't throttle properly the requests.

I missed that point. I'll remove it.